src/location/location.service.ts

```diff
@@ -0,0 +1,86 @@
+import { CacheService } from '../cache/cache.service';
+import { InjectModel } from '@nestjs/mongoose';
+import { Injectable } from '@nestjs/common';
+import { Location, LocationDocument } from './schemas/location.schema'; // Import the CacheService
+import { Model } from 'mongoose';
+
+@Injectable()
+export class LocationService {
+  constructor(
+    @InjectModel(Location.name) private _locationModel: Model<LocationDocument>,
+    private _cacheService: CacheService
+  ) {}
+
+  async findLocation(postal_code: string): Promise<Location | undefined> {
+    return this._locationModel.findOne({ PostalCode: postal_code });
+  }
+
+  async getAllLocation(): Promise<Location[]> {
+    return this._locationModel.find().exec();
+  }
+
+  async getNearbyPostalCodes(
+    distance: number,
+    postalCode: string,
+  ): Promise<Location[]> {
+    const cacheKey = `nearbyPostalCodes:${postalCode}:${distance}`;
+
+    const cachedResult = await this._cacheService.get(cacheKey);
+    if (cachedResult) {
+      return cachedResult;
+    }
+
+    const targetLocation = await this._locationModel.findOne({ PostalCode : postalCode }).exec();
+
+    if (!targetLocation) {
+      return []; // if location not found, return blank;
+    }
+
+    const allLocations = await this._locationModel.find().exec();
```

Comment:
Can put filter into query, so no need to run `.filter`
`const allLocations = await this._locationModel.find({PostalCode: {$ne: postalCode}}).exec();`

Reply:
Change query from find and fiilter -> find with exclusion